meta/include/gnuradio-4.0/meta/UncertainValue.hpp

```diff
@@ -0,0 +1,331 @@
+#ifndef GNURADIO_UNCERTAINVALUE_HPP
+#define GNURADIO_UNCERTAINVALUE_HPP
+
+#include <atomic>
+#include <complex>
+#include <concepts>
+#include <cstdint>
+#include <numbers>
+#include <optional>
+#include <type_traits>
+
+#include <gnuradio-4.0/meta/utils.hpp>
+
+namespace gr::meta {
+
+/**
+ *
+ * @brief Propagation of Uncertainties
+ *
+ * original idea by: Evan Manning, "Uncertainty Propagation in C++", NASA Jet Propulsion Laboratory,
+ * C/C++ Users Journal Volume 14, Number 3, March, 1996
+ * http://www.pennelynn.com/Documents/CUJ/HTML/14.03/MANNING/MANNING.HTM
+ *
+ * implements +,-,*,/ operators for basic arithmetic and complex types, for details see:
+ * https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae
+ * This implements only propagation of errors between variables that are uncorrelated.
+ * A more rigorous treatment would require the calculation and propagation of the
+ * corresponding covariance matrix which is out of scope of this implementation.
+ */
+
+template<typename T>
+concept arithmetic_or_complex_like = std::is_arithmetic_v<T> || complex_like<T>;
+
+template<arithmetic_or_complex_like T>
+struct UncertainValue {
+    using value_type = T;
+
+    T value       = static_cast<T>(0); /// mean value
+    T uncertainty = static_cast<T>(0); /// uncorrelated standard deviation
```

Comment:
I wish I had – I'm really not a physicist with a use case for this. Yes, we do commonly deal with I/Q imbalance, have to scale one of the branches to fit the signal amplitude of the other, and thus have different noise variances in real and imaginary part:

It's just that if you treat them as independent errors, that goes southwards as soon as you do *any* operation that "rotates" the complex value; essentially, any multiplicative operation means that afterwards, your errors in I and Q are correlated, as explained somewhere further below :) And among the first things you have to do when you have complex baseband is to rotate the phase so that the complex signal-contained information is recoverable.

Reply:
See especially [this](https://github.com/fair-acc/graph-prototype/pull/226#discussion_r1409167020) comment below; your multiplication operation, as far as I read it, keeps the output I uncertainty independent from the inputs' Q uncertainty. That's impossible.

Reply:
Dear  @marcusmueller,

Thank you for your detailed comments, time, and effort in reviewing the implementation of `UncertainValue` for complex numbers. I want to clarify the assumptions and practical applications of this code.

The PR and code description clearly outline the implementation's basis and the propagation of uncorrelated symmetric errors, specifically Gaussian-type standard deviations. 
```cpp
/**
 *
 * @brief Propagation of Uncertainties
 *
 * original idea by: Evan Manning, "Uncertainty Propagation in C++", NASA Jet Propulsion Laboratory,
 * C/C++ Users Journal Volume 14, Number 3, March, 1996
 * http://www.pennelynn.com/Documents/CUJ/HTML/14.03/MANNING/MANNING.HTM
 *
 * implements +,-,*,/ operators for basic arithmetic and complex types, for details see:
 * https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae
 * This implements only propagation of uncorrelated symmetric errors (i.e. gaussian-type standard deviations).
 * A more rigorous treatment would require the calculation and propagation of the
 * corresponding covariance matrix which is out of scope of this implementation.
 */
```

This approach is tailored to our specific use case, where we map pairs of real-valued ADC samples with real-valued, uncorrelated uncertainties for each channel to `std::complex<T>`. In our context, complex numbers serve more as a convenient representation for pairs of independent values (akin to vectors or tuples) rather than traditional complex numbers with correlated real and imaginary parts.

In our specific applications, such as gain scheduling and signal conditioning in scientific and engineering contexts, we primarily deal with linear operations where the independence of uncertainties remains valid. This includes scenarios where complex numbers are added, subtracted or scaled by non-complex numbers, and the uncertainties of each component are independent, originating from different noise sources. For multiplicative or non-linear operations, while the post-operation correlation between real and imaginary parts might occur, our focus is on the initial step where the noise sources are uncorrelated. As an 80/20 engineering approximation, this holds for most relevant use cases in signal processing and feedback control applications.

Furthermore, in the context of operations like FFT, the linear treatment of uncertainties remains applicable, e.g.
$$\mathcal{FFT}(A(t)+B(t)) = \mathcal{FFT}(A(t)) + \mathcal{FFT}(B(t))$$ and notably
$$\mathcal{FFT}(\overline{x}(t) \pm \overline{\sigma_A}(t)) = \mathcal{FFT}((\overline{x}(t)) \pm FFT(\overline{\sigma_A}(t)).$$

This linear approach simplifies the otherwise complex task of handling large covariance matrices, which, while more rigorous, would be impractical and unnecessary for the most common use cases.

I acknowledge that this implementation does not cover all possible scenarios, especially those involving correlated uncertainties or non-linear operations. However, for our targeted applications (filtering, FFT, gain-scheduling, signal conditioning, ...), this approach balances accuracy and computational feasibility. 

For practical engineering purposes, at worst, this approach slightly underestimates the propagated error compared to a fully (anti-)correlated treatment — by only about 30%, not orders of magnitude. I feel this level of approximation is should be acceptable for most use cases. Moreover, it's crucial to emphasize that having an uncertainty estimate, even if not perfect, is significantly better than having none, which is currently the status quo in GR.

I hope this helps clear doubts and motivates the reasons behind the specific implementation choices. I am open to further discussion and am eager to hear any additional new thoughts or suggestions you might have.

Reply:
It remains wrong, and in a worse way, for the multiplication of two complexes with uncertainties, because that isn't even a linear operation.

However, when you look at your source code for the uncertainty of the imaginary part of the product, you see that the input's uncertainty in the real part doesn't seem to play a role. There's just some fundamentally wrong math here, and I don't really understand where the math comes from, so that's why I keep asking for an actual reference rather than just handwaving.